src/LinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -214,52 +217,93 @@ UprobesUnwindingVisitor::ComputeCallstackTypeFromStackSample(
       stack_slices.emplace_back(user_stack_slice.start_address, user_stack_slice.size,
                                 user_stack_slice.data.get());
     }
   }
 
-  LibunwindstackResult libunwindstack_result = unwinder_->Unwind(
-      event_data.pid, current_maps_->Get(), event_data.GetRegistersAsArray(), stack_slices);
+  LibunwindstackResult libunwindstack_result =
+      unwinder_->Unwind(event_data.GetCallstackPidOrMinusOne(), current_maps_->Get(),
+                        event_data.GetRegistersAsArray(), stack_slices);
 
   if (libunwindstack_result.frames().empty()) {
     // Even with unwinding errors this is not expected because we should at least get the program
     // counter. Do nothing in case this doesn't hold for a reason we don't know.
     ORBIT_ERROR("Unwound callstack has no frames");
-    return;
+    return false;
   }
 
-  FullCallstackSample sample;
-  sample.set_pid(event_data.pid);
-  sample.set_tid(event_data.tid);
-  sample.set_timestamp_ns(event_timestamp);
-
-  Callstack* callstack = sample.mutable_callstack();
   callstack->set_type(ComputeCallstackTypeFromStackSample(libunwindstack_result));
   for (const unwindstack::FrameData& libunwindstack_frame : libunwindstack_result.frames()) {
     SendFullAddressInfoToListener(libunwindstack_frame);
     callstack->add_pcs(libunwindstack_frame.pc);
   }
 
   ORBIT_CHECK(!callstack->pcs().empty());
+  return true;
+}
+
+void UprobesUnwindingVisitor::Visit(uint64_t event_timestamp,
+                                    const StackSamplePerfEventData& event_data) {
+  FullCallstackSample sample;
+  sample.set_pid(event_data.pid);
+  sample.set_tid(event_data.tid);
+  sample.set_timestamp_ns(event_timestamp);
+
+  const bool success = UnwindStack(event_data, sample.mutable_callstack());
+
+  if (!success) {
+    return;
+  }
+
   listener_->OnCallstackSample(std::move(sample));
 }
 
+void UprobesUnwindingVisitor::Visit(uint64_t event_timestamp,
+                                    const SchedWakeupWithStackPerfEventData& event_data) {
+  ThreadStateSliceCallstack thread_state_slice_callstack;
+  thread_state_slice_callstack.set_thread_state_slice_tid(event_data.woken_tid);
+  thread_state_slice_callstack.set_timestamp_ns(event_timestamp);
+
+  const bool success = UnwindStack(event_data, thread_state_slice_callstack.mutable_callstack());
+
+  if (!success) {
+    return;
+  }
+
+  listener_->OnThreadStateSliceCallstack(std::move(thread_state_slice_callstack));
+}
+
+void UprobesUnwindingVisitor::Visit(uint64_t event_timestamp,
+                                    const SchedSwitchWithStackPerfEventData& event_data) {
+  ThreadStateSliceCallstack thread_state_slice_callstack;
+  thread_state_slice_callstack.set_thread_state_slice_tid(event_data.prev_tid);
+  thread_state_slice_callstack.set_timestamp_ns(event_timestamp);
+  bool const success = UnwindStack(event_data, thread_state_slice_callstack.mutable_callstack());
```

Comment:
```suggestion
  thread_state_slice_callstack.set_timestamp_ns(event_timestamp);

  bool const success = UnwindStack(event_data, thread_state_slice_callstack.mutable_callstack());
```
You have it in the other two.

Reply:
Done.